packages/functionals/botpress-qna/src/providers/microsoft_qna_maker.js

```diff
@@ -0,0 +1,171 @@
+import _ from 'lodash'
+import axios from 'axios'
+import ms from 'ms'
+import Promise from 'bluebird'
+
+const KNOWLEDGEBASE_NAME = 'botpress'
+
+// Handles QnA Maker API downcasing all key-values in metadata
+const markUpperCase = str => str.replace(/([A-Z])/g, 'sssooo$1sss')
+const restoreUpperCase = str =>
+  str
+    .split('sss')
+    .map(chunk => (chunk.startsWith('ooo') ? chunk.slice(3).toUpperCase() : chunk))
+    .join('')
+const keysToRestore = { redirectflow: 'redirectFlow', redirectnode: 'redirectNode' }
+
+const qnaItemData = ({ questions, answer, metadata }) => ({
+  questions,
+  answer,
+  ..._.fromPairs(metadata.map(({ name, value }) => [keysToRestore[name] || name, restoreUpperCase(value)])),
+  enabled: (metadata.find(({ name }) => name === 'enabled') || {}).value === 'true'
+})
+
+const prepareMeta = data =>
+  _.chain(data)
+    .pick(['enabled', 'action', 'redirectFlow', 'redirectNode'])
+    .toPairs()
+    .map(([name, value]) => ({ name, value: _.isString(value) ? markUpperCase(value) : value }))
+    .filter(({ value }) => !_.isUndefined(value) && value !== '')
+    .value()
+
+export default class Storage {
+  constructor({ bp, config }) {
+    const baseURL = 'https://westus.api.cognitive.microsoft.com/qnamaker/v4.0'
+    const headers = { 'Ocp-Apim-Subscription-Key': config.microsoftQnaMakerApiKey }
+    Object.assign(this, { bp, client: axios.create({ baseURL, headers }) })
+  }
+
+  async initialize() {
+    const isBpKnowledgbase = ({ name }) => name === KNOWLEDGEBASE_NAME
+    const { data: { knowledgebases: initialKnowledgebases } } = await this.client.get('/knowledgebases/')
+    const existingKb = initialKnowledgebases.find(isBpKnowledgbase)
+    if (existingKb) {
+      this.knowledgebase = existingKb
+    } else {
+      const { data: { operationId } } = await this.client.post('/knowledgebases/create', { name: KNOWLEDGEBASE_NAME })
+      await this.waitForOperationToFinish(operationId)
+      const { data: { knowledgebases } } = await this.client.get('/knowledgebases/')
+      this.knowledgebase = knowledgebases.find(isBpKnowledgbase)
+    }
+
+    this.endpointKey = (await this.client.get('/endpointkeys')).data.primaryEndpointKey
+  }
+
+  publish = () => this.client.post(`/knowledgebases/${this.knowledgebase.id}`)
+
+  patchKb = params => this.client.patch(`/knowledgebases/${this.knowledgebase.id}`, params)
+
+  waitForOperationToFinish = async operationId => {
+    await Promise.delay(ms(200))
+    while (true) {
+      const { data, headers: { 'retry-after': timeout } } = await this.client.get(`/operations/${operationId}`)
```

Comment:
Where is `timeout` coming from? I don't see it defined

Reply:
It's getting assigned here. I'm not using it below since they return large timeouts with 30 seconds which doesn't usually make sense.

Reply:
Sorry I've misread this line